<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Исследование-надежности-заемщиков" data-toc-modified-id="Исследование-надежности-заемщиков-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Исследование надежности заемщиков</a></span><ul class="toc-item"><li><span><a href="#Импорт-данных-и-общая-информация" data-toc-modified-id="Импорт-данных-и-общая-информация-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Импорт данных и общая информация</a></span></li></ul></li><li><span><a href="#Предобработка-данных" data-toc-modified-id="Предобработка-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Предобработка данных</a></span><ul class="toc-item"><li><span><a href="#Удаление-пропусков" data-toc-modified-id="Удаление-пропусков-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Удаление пропусков</a></span></li><li><span><a href="#Обработка-аномальных-значений" data-toc-modified-id="Обработка-аномальных-значений-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Обработка аномальных значений</a></span></li><li><span><a href="#Удаление-пропусков-(продолжение)" data-toc-modified-id="Удаление-пропусков-(продолжение)-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Удаление пропусков (продолжение)</a></span></li><li><span><a href="#Изменение-типов-данных" data-toc-modified-id="Изменение-типов-данных-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Изменение типов данных</a></span></li><li><span><a href="#Обработка-дубликатов" data-toc-modified-id="Обработка-дубликатов-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Обработка дубликатов</a></span></li><li><span><a href="#Категоризация-данных" data-toc-modified-id="Категоризация-данных-2.6"><span class="toc-item-num">2.6&nbsp;&nbsp;</span>Категоризация данных</a></span></li></ul></li><li><span><a href="#Исследование-данных" data-toc-modified-id="Исследование-данных-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Исследование данных</a></span><ul class="toc-item"><li><span><a href="#Есть-ли-зависимость-между-количеством-детей-и-возвратом-кредита-в-срок?" data-toc-modified-id="Есть-ли-зависимость-между-количеством-детей-и-возвратом-кредита-в-срок?-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Есть ли зависимость между количеством детей и возвратом кредита в срок?</a></span></li><li><span><a href="#Есть-ли-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок?" data-toc-modified-id="Есть-ли-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок?-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Есть ли зависимость между семейным положением и возвратом кредита в срок?</a></span></li><li><span><a href="#Есть-ли-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок?" data-toc-modified-id="Есть-ли-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок?-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Есть ли зависимость между уровнем дохода и возвратом кредита в срок?</a></span></li><li><span><a href="#Как-разные-цели-кредита-влияют-на-его-возврат-в-срок?" data-toc-modified-id="Как-разные-цели-кредита-влияют-на-его-возврат-в-срок?-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Как разные цели кредита влияют на его возврат в срок?</a></span></li><li><span><a href="#Возможные-причины-появления-пропусков-в-исходных-данных." data-toc-modified-id="Возможные-причины-появления-пропусков-в-исходных-данных.-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>Возможные причины появления пропусков в исходных данных.</a></span></li><li><span><a href="#Почему-заполнить-пропуски-медианным-значением-—-лучшее-решение-для-количественных-переменных." data-toc-modified-id="Почему-заполнить-пропуски-медианным-значением-—-лучшее-решение-для-количественных-переменных.-3.6"><span class="toc-item-num">3.6&nbsp;&nbsp;</span>Почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.</a></span></li></ul></li><li><span><a href="#Общий-вывод." data-toc-modified-id="Общий-вывод.-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Общий вывод.</a></span></li></ul></div>

## Исследование надежности заемщиков

### Импорт данных и общая информация

In [1]:
import numpy as np
import pandas as pd
import warnings
warnings.filterwarnings('ignore')

In [2]:
try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('...')

In [3]:
data

children  days_employed  dob_years education  education_id  \
0             1   -8437.673028         42    высшее             0   
1             1   -4024.803754         36   среднее             1   
2             0   -5623.422610         33   Среднее             1   
3             3   -4124.747207         32   среднее             1   
4             0  340266.072047         53   среднее             1   
...         ...            ...        ...       ...           ...   
21520         1   -4529.316663         43   среднее             1   
21521         0  343937.404131         67   среднее             1   
21522         1   -2113.346888         38   среднее             1   
21523         3   -3112.481705         38   среднее             1   
21524         2   -1984.507589         40   среднее             1   

          family_status  family_status_id gender income_type  debt  \
0       женат / замужем                 0      F   сотрудник     0   
1       женат / замужем                 0      F   сотрудник     0   
2       женат / замужем                 0      M   сотрудник     0   
3       женат / замужем                 0      M   сотрудник     0   
4      гражданский брак                 1      F   пенсионер     0   
...                 ...               ...    ...         ...   ...   
21520  гражданский брак                 1      F   компаньон     0   
21521   женат / замужем                 0      F   пенсионер     0   
21522  гражданский брак                 1      M   сотрудник     1   
21523   женат / замужем                 0      M   сотрудник     1   
21524   женат / замужем                 0      F   сотрудник     0   

        total_income                       purpose  
0      253875.639453                 покупка жилья  
1      112080.014102       приобретение автомобиля  
2      145885.952297                 покупка жилья  
3      267628.550329    дополнительное образование  
4      158616.077870               сыграть свадьбу  
...              ...                           ...  
21520  224791.862382             операции с жильем  
21521  155999.806512          сделка с автомобилем  
21522   89672.561153                  недвижимость  
21523  244093.050500  на покупку своего автомобиля  
21524   82047.418899         на покупку автомобиля  

[21525 rows x 12 columns]

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

In [5]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

In [6]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

In [7]:
data['days_employed'] = data['days_employed'].abs()

In [8]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

In [9]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5], dtype=int64)

In [10]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

In [11]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5], dtype=int64)

### Удаление пропусков (продолжение)

In [12]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

In [13]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

In [14]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

In [15]:
data['education'] = data['education'].str.lower()

In [16]:
data.duplicated().sum()

71

In [17]:
data = data.drop_duplicates()

### Категоризация данных

In [18]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [19]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

In [20]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

In [21]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [22]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

## Исследование данных

### Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [23]:
data.groupby('children').debt.mean()

children
0    0.075438
1    0.092346
2    0.094542
3    0.081818
4    0.097561
5    0.000000
Name: debt, dtype: float64

Вычисляю части должников от всех заемщиков в зависимости от количества детей. Зависимости вида "чем больше детей, тем большая часть должников" не наблюдается, но из данных видно, что часть должников среди бездетных заемщиков несколько ниже, чем среди детных.

In [24]:
data.groupby('children').debt.count()

children
0    14091
1     4808
2     2052
3      330
4       41
5        9
Name: debt, dtype: int64

Рассмотрим количество заемщиков для каждой категории. Заемщиков с 5, 4, 3 детьми не так много в выборке.

In [25]:
sum(data['children'].isin((3, 4, 5))) / len(data)

0.017814448455299798

Всего лишь около 1,8% от всех заемщиков. Объединим в одну категорию многодетных заемщиков, имеющих 2 и более детей и вычислим часть должников для трех категорий: бездетные, с одним ребенком и многодетные.

In [26]:
data[data['children'] == 0]['debt'].mean() # бездетные

0.07543822297920659

In [27]:
data[data['children'] == 1]['debt'].mean() # с одним ребенком

0.09234608985024959

In [28]:
data[data['children'].isin((2, 3, 4, 5))]['debt'].mean() # многодетные

0.09251644736842106

In [29]:
data[data['children'] != 0]['debt'].mean() - data[data['children'] == 0]['debt'].mean() # разница между детными и бездетными

0.01696509193792048

Категории с одним ребенком и многодетные показывают практически одинаковую часть должников (0.0923 и 0.0925), тогда как бездетные показывают более низкую часть должников (0.0754). Разница в частях должников между детными и бездетными 0.017.

**Вывод:** Зависимости между количеством детей и возвратом кредита нет. Зависимость между наличием (независимо от количества) либо отсутствием детей и возвратом кредита есть.

### Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [30]:
data.groupby('family_status')['debt'].agg(number_of_debtors='sum', number_of_borrowers='count', part_of_debtors='mean')\
.reset_index()

family_status  number_of_debtors  number_of_borrowers  \
0  Не женат / не замужем                273                 2796   
1              в разводе                 84                 1189   
2         вдовец / вдова                 63                  951   
3       гражданский брак                385                 4134   
4        женат / замужем                927                12261   

   part_of_debtors  
0         0.097639  
1         0.070648  
2         0.066246  
3         0.093130  
4         0.075606

Наблюдается зависимость возврата кредита от того находился ли когда-либо заемщик в законном браке (в разводе, вдовец / вдова, женат / замужем) или нет (гражданский брак, Не женат / не замужем). Заемщики бывашие в законных отношениях показывают более низкие части должников.

In [31]:
data[data['family_status'].isin(['в разводе', 'вдовец / вдова', 'женат / замужем'])]['debt'].mean() # были в законных отношениях

0.07457815429484063

In [32]:
data[data['family_status'].isin(['гражданский брак', 'Не женат / не замужем'])]['debt'].mean() # не были в законных отношениях

0.09494949494949495

In [33]:
0.09494949494949495 - 0.07457815429484063

0.02037134065465432

Разница в частях должников между не бывашими в законном браке и бывавшими 0.02.

**Вывод:** Зависимость между наличаем (в настоящем или в прошлом) законных семейных отношений и возвратом кредита есть. Заемщики бывашие в законных отношениях показывают более низкие части должников.

### Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [34]:
data.groupby('total_income_category').debt.mean()

total_income_category
A    0.080000
B    0.070602
C    0.084982
D    0.060172
E    0.090909
Name: debt, dtype: float64

Уровни дохода расположены от самого высокого (A) до самого низкого (E). Не наблюдается ни монотонного возрастания, ни монотонного убывания части должников от категории А к категории Е.

In [35]:
data.total_income_category.value_counts()

C    15921
B     5014
D      349
A       25
E       22
Name: total_income_category, dtype: int64

**Вывод:** Зависимости между уровнем дохода и возвратом кредита нет.

### Как разные цели кредита влияют на его возврат в срок?

In [36]:
data.pivot_table(values='debt', index='purpose_category', aggfunc=[len, np.mean]).reset_index()

purpose_category    len      mean
                              debt      debt
0    операции с автомобилем   4279  0.093480
1  операции с недвижимостью  10751  0.072551
2     получение образования   3988  0.092528
3        проведение свадьбы   2313  0.079118

**Вывод:** Цели кредита влияют на его возврат. Самые "опасные" цели для невозврата связаны с автомобилями, самые "безопасные" - с недвижимостью.

### Возможные причины появления пропусков в исходных данных.

Либо человеческий фактор заемщих либо оператор пропустил пунк призаполнении заявки (возможно случайно возможно из-за желания скрыть некоторые факты о себе), либо произошел сбой при выгрузки данных файл из базы данных.

### Почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

Медиана лучшее решение для заполнения пропусков, потому что медиана(в отличае от среднего значения) малочувствительна к выбросам (например, в датасете может быть заемщик с доходом, значительно превышающим доходы большинства заемщиков) и ошибочно введеным данных (оператор мог случайно ввести отрицательное число вместо положительного или перепутать единицы измерения). 

## Общий вывод.

1. Есть отрицательная зависимость между наличием детей (независимо от количества) и возвратом кредита.
2. Есть положительная зависимость между наличаем (в настоящем или в прошлом) законных семейных отношений и возвратом кредита есть.
3. Цели кредита влияют на возврат кредита.Самые "опасные" цели для невозврата связаны с автомобилями, самые "безопасные" - с недвижимостью.
4. Зависимостей между уровнем дохода и возвратом кредита не выявлено.